torch/csrc/jit/codegen/cuda/root_domain_map.cpp

```diff
@@ -0,0 +1,379 @@
+#include <torch/csrc/jit/codegen/cuda/root_domain_map.h>
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/iter_visitor.h>
+
+#include <sstream>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+bool hasMatchingDomains(const std::vector<DomainKey>& domains,
+                        const DisjointSet<DomainKey, DomainKeyHash>& eq_set) {
+  for (const auto& key: domains) {
+    for (const auto& other_key: domains) {
+      if (key == other_key) continue;
+      const auto& root = other_key.td()->getRootDomain();
+      if (std::any_of(root.begin(), root.end(),
+                      [&](const IterDomain* id) {
+                        return eq_set.areEquivalent(key, DomainKey(other_key.td(), id));
+                      })) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool safeToJoin(const std::unordered_set<DomainKey, DomainKeyHash>& domains,
+                const DisjointSet<DomainKey, DomainKeyHash>& eq_set,
+                const FindIncompatibleDomains& inconsistent_domains) {
+  //std::cerr << "safeTojoin?: " << domains << std::endl;
+  if (domains.size() <= 1) {
+    return true;
+  }
+  // Filter out equivalent domains
+  std::vector<DomainKey> unique_domains;
+  for (const auto& domain: domains) {
+    if (std::none_of(unique_domains.begin(), unique_domains.end(),
+                     [&](const auto& unique_dom) {
+                       return eq_set.areEquivalent(domain, unique_dom);
+                     })) {
+      unique_domains.push_back(domain);
+    }
+  }
+  //std::cerr << "Consumer domains: " << unique_domains << std::endl;
+  if (hasMatchingDomains(unique_domains, eq_set)) {
+    //std::cerr << "Has matching domains" << std::endl;
+    return false;
+  }
+  if (inconsistent_domains.isReductionOutputMerged(unique_domains, eq_set)) {
+    //std::cerr << "Prevented by reduction" << std::endl;
+    return false;
+  }
+  return true;
+}
+
+} // namespace
+
+std::ostream& DomainKey::print(std::ostream& os) const {
+  std::stringstream ss;
+  ss << "{";
+  if (td_) {
+    ss << td_ << " (root: " << td_->getRootDomain()
+       << ", maybe rfactor: " << td_->getMaybeRFactorDomain() << ")";
+  } else {
+    ss << "null";
+  }
+  ss << ", ";
+  if (id_) {
+    ss << id_;
+  } else {
+    ss << "null";
+  }
+  ss << "}";
+  return os << ss.str();
+}
+
+FindIncompatibleDomains::FindIncompatibleDomains() {
+  Fusion* fusion = FusionGuard::getCurFusion();
+  traverse(fusion);
+}
+
+void FindIncompatibleDomains::handle(ReductionOp* op) {
```

Comment:
Map any reduction IterDomains as incompatible with all IterDomains in all consumer domains.